File: app/src/main/java/com/recursia/popularmovies/presentation/views/fragments/SearchScreenFragment.java
package com.recursia.popularmovies.presentation.views.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.recursia.popularmovies.R;
import com.recursia.popularmovies.TheApplication;
import com.recursia.popularmovies.di.AppComponent;
import com.recursia.popularmovies.domain.models.Movie;
import com.recursia.popularmovies.presentation.presenters.SearchScreenPresenter;
import com.recursia.popularmovies.presentation.views.adapters.MoviesAdapter;
import com.recursia.popularmovies.presentation.views.contracts.SearchScreenContract;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SearchScreenFragment extends MvpAppCompatFragment implements SearchScreenContract {
    private static final int SPAN_COUNT = 2;

    @BindView(R.id.searchRecyclerView)
    RecyclerView searchRecyclerView;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @InjectPresenter
    SearchScreenPresenter presenter;
    private MoviesAdapter moviesAdapter;

    public static SearchScreenFragment getInstance() {
        return new SearchScreenFragment();
    }

    @Override
    public void showErrorMessage(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
    }

    @ProvidePresenter
    SearchScreenPresenter providePresenter() {
        AppComponent app = TheApplication.getInstance().getAppComponent();
        return new SearchScreenPresenter(app.getSearchScreenInteractor(), app.getRouter());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_search, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initToolbar();
        initRecyclerView();
        initAdapter();
    }

    private void initToolbar() {
        toolbar.setTitle(getString(R.string.search_movie_title));
        toolbar.setBackgroundColor(getResources().getColor(R.color.black));
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back);
        toolbar.setNavigationOnClickListener(v -> presenter.onBackPressed());
        toolbar.inflateMenu(R.menu.search_menu);
        MenuItem searchItem = toolbar.getMenu().findItem(R.id.searchItem);
        SearchView searchView = (SearchView) searchItem.getActionView();

        searchView.setQueryHint(getString(R.string.search_hint));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                presenter.onQueryTextChanged(s);
                return false;
            }
        });
    }

    private void initRecyclerView() {
        searchRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), SPAN_COUNT));
        searchRecyclerView.setHasFixedSize(true);
    }

    private void initAdapter() {
        moviesAdapter = new MoviesAdapter(getContext(), false);
        moviesAdapter.setClickListener(item -> presenter.onItemClicked(item));
        searchRecyclerView.setAdapter(moviesAdapter);
    }

    @Override
    public void setMovies(List<Movie> movies) {
        moviesAdapter.setMovies(movies);
    }

    @Override
    public void addMovies(List<Movie> movies) {
        moviesAdapter.addMovies(movies);
    }

}

File: app/src/main/java/com/recursia/popularmovies/presentation/views/fragments/SearchScreenFragment.kt
package com.recursia.popularmovies.presentation.views.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.recursia.popularmovies.R
import com.recursia.popularmovies.TheApplication
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.presentation.presenters.SearchScreenPresenter
import com.recursia.popularmovies.presentation.views.adapters.MoviesAdapter
import com.recursia.popularmovies.presentation.views.adapters.common.MovieStretchItemType
import com.recursia.popularmovies.presentation.views.contracts.SearchScreenContract
import com.recursia.popularmovies.presentation.views.decorations.MarginItemDecoration
import moxy.MvpAppCompatFragment
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter

class SearchScreenFragment : MvpAppCompatFragment(), SearchScreenContract {
    @BindView(R.id.searchRecyclerView)
    lateinit var searchRecyclerView: RecyclerView

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @InjectPresenter
    lateinit var presenter: SearchScreenPresenter
    private lateinit var moviesAdapter: MoviesAdapter
    override fun showErrorMessage(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    @ProvidePresenter
    fun providePresenter(): SearchScreenPresenter {
        val app = TheApplication.getInstance().appComponent
        return SearchScreenPresenter(app.searchScreenInteractor, app.router)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_search, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initToolbar()
        initRecyclerView()
        initAdapter()
    }

    private fun initToolbar() {
        toolbar.title = getString(R.string.search_movie_title)
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back)
        toolbar.setNavigationOnClickListener { presenter.onBackPressed() }
        toolbar.setBackgroundColor(resources.getColor(R.color.colorPrimary))
        toolbar.inflateMenu(R.menu.search_menu)
        val searchItem = toolbar.menu.findItem(R.id.search_item)
        val searchView = searchItem.actionView as SearchView
        searchItem.expandActionView()
        searchView.queryHint = getString(R.string.search_hint)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                presenter.onQueryTextChanged(s)
                return false
            }
        })
    }

    private fun initRecyclerView() {
        searchRecyclerView.layoutManager = GridLayoutManager(context, SPAN_COUNT)
        searchRecyclerView.addItemDecoration(
                MarginItemDecoration(context!!, 10, 10, 5, 5)
        )
        searchRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                val isBottomReached = !recyclerView.canScrollVertically(DIRECTION_DOWN)
                if (isBottomReached) {
                    presenter.bottomIsReached()
                }
            }
        })
        searchRecyclerView.setHasFixedSize(true)
    }

    private fun initAdapter() {
        moviesAdapter = MoviesAdapter(context!!, MovieStretchItemType())
        moviesAdapter.setOnClickListener {
            presenter.onItemClicked(it)
        }
        searchRecyclerView.adapter = moviesAdapter
    }

    override fun setMovies(movies: List<Movie>) {
        moviesAdapter.setMovies(movies as MutableList<Movie>)
    }

    override fun addMovies(movies: List<Movie>) {
        moviesAdapter.addMovies(movies)
    }

    companion object {
        private const val SPAN_COUNT = 2
        private const val DIRECTION_DOWN = 1
        val instance: SearchScreenFragment
            get() = SearchScreenFragment()
    }
}
